velox/exec/fuzzer/JoinMaker.h

```diff
@@ -0,0 +1,271 @@
+/// Represents a single input "table" to a join plan created using JoinMaker.
+/// This can be converted to a ValuesNode or a TableScanNode in the resulting
+/// plan.
+struct JoinSource {
+ public:
+  /// @param outputType The RowType the source node will eventually produce,
+  /// this should match the Types of the Vectors in `batches`.
+  /// @param keys The names of the key fields in `outputType`.
+  /// @param batches The data this source will produce.
+  /// @param projections Projections to be applied to batches after they are
+  /// output.
+  /// @param splitsDir A temporary directory to write files containing `batches`
+  /// into when generating splits for a TableScan.
+  /// @param writerPool A MemoryPool that can be used by the Writers when
+  /// writing out splits.
+  JoinSource(
+      const RowTypePtr& outputType,
+      const std::vector<std::string>& keys,
+      const std::vector<RowVectorPtr>& batches,
+      const std::vector<core::ExprPtr>& projections,
+      const std::string& splitsDir,
+      const std::shared_ptr<memory::MemoryPool> writerPool)
+      : outputType_(outputType),
+        keys_(keys),
+        batches_(batches),
+        projections_(projections),
+        splitsDir_(splitsDir),
+        writerPool_(writerPool) {}
+
+  const RowTypePtr& outputType() const {
+    return outputType_;
+  }
+
+  const std::vector<std::string>& keys() const {
+    return keys_;
+  }
+
+  const std::vector<RowVectorPtr>& batches() const {
+    return batches_;
+  }
+
+  /// Get `batches` as flat Vectors.
+  const std::vector<RowVectorPtr>& flatBatches() const;
+
+  /// Get `batches` as splits.
+  const std::vector<Split>& splits() const;
+
+  /// Get `batches` as splits grouped into `numGroups` groups.
+  const std::vector<Split>& groupedSplits(const int32_t numGroups) const;
+
+  const std::vector<core::ExprPtr>& projections() const {
+    return projections_;
+  }
+
+ private:
+  std::vector<RowVectorPtr> flatten(
+      const std::vector<RowVectorPtr>& vectors) const {
+    std::vector<RowVectorPtr> flatVectors;
+    for (const auto& vector : vectors) {
+      auto flat = BaseVector::create<RowVector>(
+          vector->type(), vector->size(), vector->pool());
+      flat->copy(vector.get(), 0, 0, vector->size());
+      flatVectors.push_back(flat);
+    }
+
+    return flatVectors;
+  }
+
+  std::vector<std::vector<RowVectorPtr>> splitInputByGroup(
+      int32_t numGroups) const;
+
+  const RowTypePtr outputType_;
+  const std::vector<std::string> keys_;
+  const std::vector<RowVectorPtr> batches_;
+  const std::vector<core::ExprPtr> projections_;
+  const std::string splitsDir_;
+  const std::shared_ptr<memory::MemoryPool> writerPool_;
+
+  // The following are used to cache `batches_` when it is converted to other
+  // forms.
+  mutable std::optional<std::vector<RowVectorPtr>> flatBatches_;
+  mutable std::optional<std::vector<Split>> splits_;
+  mutable std::unordered_map<int32_t, std::vector<Split>> groupedSplits_;
```

Comment:
Why would we prefer to mark them `mutable` instead of making functions changing them be non-const? Especially splits_ and groupedSplits_.

Reply:
The core state of the class is immutable and I wanted the code to reflect that, these just hold alternate representations of batches_ that are cached so we don't have to recreate them every time their respective getter methods are called. This is an implementation detail of the class that I didn't want to expose in the external API.